### PR TITLE
[Merged by Bors] - feat(ModelTheory): lift of equivalences between direct limits, and direct limit of system of substructures

### PR DESCRIPTION
**Pull request by GabinKolly**
Define the equivalence between direct limits of isomorphic systems, and the equivalence between the direct limit of a system of substructures and the union of these substructures.
Co-authored-by: Aaron Anderson <awainverse@gmail.com>

---
Important additions:
- In DirectLimit:
  - Defined `Equiv_iSup`, the isomorphism between a direct limit of a system of substructures and their union.
  - Defined `equiv_lift`, the equivalence between the direct limits of two isomorphic directed systems
  
Also added some results about limits, and simple simp lemmas.
This is some preparation work for  #9967.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 1]
```diff
@@ -391,6 +413,43 @@ theorem lift_unique (F : DirectLimit G f ↪[L] P) (x) :
...
+
+variable (L ι G f)
+
+variable (G' : ι → Type w') [∀ i, L.Structure (G' i)]
+
+variable (f' : ∀ i j, i ≤ j → G' i ↪[L] G' j)
+
+variable (g : ∀ i, G i ≃[L] G' i)
+
+variable (H_commutating : ∀ i j hij x, g j (f i j hij x) = f' i j hij (g i x))
```
Comment (awainverse) [inline Mathlib/ModelTheory/DirectLimit.lean:430]:
```suggestion
variable (H_commuting : ∀ i j hij x, g j (f i j hij x) = f' i j hij (g i x))
```

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 2]
```diff
@@ -391,6 +413,43 @@ theorem lift_unique (F : DirectLimit G f ↪[L] P) (x) :
...
+  have surj_f : Function.Surjective F := by
+    intro x
+    rcases x with ⟨i, pre_x⟩
+    use of L _ G f i ((g i).symm pre_x)
+    simp only [lift_of, Embedding.comp_apply, Equiv.coe_toEmbedding, Equiv.apply_symm_apply]
+    rfl
+  exact ⟨Equiv.ofBijective F ⟨F.injective, surj_f⟩, F.map_fun', F.map_rel'⟩
+
+theorem equiv_lift_of {i : ι} (x : G i) :
+    equiv_lift L ι G f G' f' g H_commutating (of L ι G f i x) = of L ι G' f' i (g i x) := rfl
```
Comment (awainverse) [inline Mathlib/ModelTheory/DirectLimit.lean:449]:
```suggestion
    equiv_lift L ι G f G' f' g H_commuting (of L ι G f i x) = of L ι G' f' i (g i x) := rfl
```

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 3]
```diff
@@ -391,6 +413,43 @@ theorem lift_unique (F : DirectLimit G f ↪[L] P) (x) :
...
+variable (H_commutating : ∀ i j hij x, g j (f i j hij x) = f' i j hij (g i x))
+
+variable [DirectedSystem G' fun i j h => f' i j h]
+
+/-- The isomorphism between limits of isomorphic systems. -/
+noncomputable def equiv_lift : DirectLimit G f ≃[L] DirectLimit G' f' := by
+  let U i : G i ↪[L] DirectLimit G' f' := (of L _ G' f' i).comp (g i).toEmbedding
+  let F : DirectLimit G f ↪[L] DirectLimit G' f' := lift L _ G f U <| by
+    intro _ _ _ _
+    simp only [Embedding.comp_apply, Equiv.coe_toEmbedding, H_commutating, of_f]
```
Comment (awainverse) [inline Mathlib/ModelTheory/DirectLimit.lean:439]:
```suggestion
    simp only [Embedding.comp_apply, Equiv.coe_toEmbedding, H_commuting, of_f]
```

**Review by awainverse — CHANGES_REQUESTED** [review]
I approve this work - I just want to correct a variable name.

File: Mathlib/ModelTheory/DirectLimit.lean [hunk 4]
```diff
@@ -391,6 +413,43 @@ theorem lift_unique (F : DirectLimit G f ↪[L] P) (x) :
...
+
+variable (G' : ι → Type w') [∀ i, L.Structure (G' i)]
+
+variable (f' : ∀ i j, i ≤ j → G' i ↪[L] G' j)
+
+variable (g : ∀ i, G i ≃[L] G' i)
+
+variable (H_commuting : ∀ i j hij x, g j (f i j hij x) = f' i j hij (g i x))
+
+variable [DirectedSystem G' fun i j h => f' i j h]
```
Comment (jcommelin) [inline Mathlib/ModelTheory/DirectLimit.lean:432]:
```suggestion
variable (L ι G f)
variable (G' : ι → Type w') [∀ i, L.Structure (G' i)]
variable (f' : ∀ i j, i ≤ j → G' i ↪[L] G' j)
variable (g : ∀ i, G i ≃[L] G' i)
variable (H_commuting : ∀ i j hij x, g j (f i j hij x) = f' i j hij (g i x))
variable [DirectedSystem G' fun i j h => f' i j h]
```

**Pull request closed without merging**